include/nanorange/algorithm/stable_partition.hpp

```diff
@@ -0,0 +1,177 @@
+
+        if (dist == 2) {
+            // We know first is false and last is true, so swap them
+            nano::iter_swap(first, last);
+            return last;
+        }
+
+        if (dist == 3) {
+            // We know first is false and last is true, so look at middle
+            I middle = nano::next(first);
+
+            if (nano::invoke(pred, nano::invoke(proj, *middle))) {
+                nano::iter_swap(first, middle);
+                nano::iter_swap(middle, last);
+                return last;
+            }
+
+            // middle is false
+            nano::iter_swap(middle, last);
+            nano::iter_swap(first, middle);
+            return middle;
+        }
+
+        const dist_t half = dist/2;
+        const I middle = nano::next(first, half);
+
+        I m1 = nano::prev(middle);
+        dist_t len_half = half;
+
+        while (m1 != first && !nano::invoke(pred, nano::invoke(proj, *m1))) {
+            --len_half;
+            --m1;
+        }
+
+        const I first_false = (m1 == first) ? first :
+                impl_unbuffered(first, m1, len_half, pred, proj);
+
+        m1 = middle;
+        len_half = dist - half;
+
+        while (nano::invoke(pred, nano::invoke(proj, *m1))) {
+            if (++m1 == last) {
+                return nano::rotate(first_false, middle, ++last).begin();
+            }
+        }
+
+        const I last_false = impl_unbuffered(m1, last, len_half, pred, proj);
+
+        return nano::rotate(first_false, middle, last_false).begin();
+    }
+
+    template <typename I, typename Pred, typename Proj>
+    static I impl(I first, I last, Pred& pred, Proj& proj)
+    {
+        // Find the first non-true value
+        first = nano::find_if_not(std::move(first), last, std::ref(pred), std::ref(proj));
+        if (first == last) {
+            return first;
+        }
+
+        // Find the last true value
+        last = nano::find_if(nano::make_reverse_iterator(last),
+                             nano::make_reverse_iterator(first),
+                             std::ref(pred), std::ref(proj)).base();
+        if (last == first) {
+            return first;
+        }
+
+        const auto dist = nano::distance(first, last);
+
+        auto buf = detail::temporary_vector<iter_value_t<I>>(dist);
+        if (buf.capacity() < static_cast<std::size_t>(dist)) {
+            return impl_unbuffered(first, --last, dist, pred, proj);
+        }
+        return impl_buffered(first, last, buf, pred, proj);
+    }
+
+    template <typename I, typename S, typename Pred, typename Proj>
+    static std::enable_if_t<!Same<I, S>, I>
+    impl(I first, S last, Pred& pred, Proj& proj)
+    {
+        return impl(first, nano::next(first, last), pred, proj);
+    }
+
+public:
+    template <typename I, typename S, typename Pred, typename Proj = identity>
+    std::enable_if_t<
+        BidirectionalIterator<I> &&
+        Sentinel<S, I> &&
+        IndirectUnaryPredicate<Pred, projected<I, Proj>>, I>
```

Comment:
These requirements look wrong; shouldn't this be `Sortable<I, Pred, Proj>`?

Reply:
The working draft has `IndirectUnaryPredicate<Pred, projected<I, Proj>>`... and `Permutable<I>`, which apparently I missed.

Reply:
I noticed `Permutable` wasn't there, and it occurred to me that `stable_sort` doesn't need `Permutable` because it gets it from `Sortable`. Were this an implementation of `stable_sort`, my suggestion would have been substantially more pertinent.